plat/mediatek/mt6795/bl31_plat_setup.c

```diff
@@ -0,0 +1,483 @@
+ * the security state specified. BL33 corresponds to the non-secure image type
+ * while BL32 corresponds to the secure image type. A NULL pointer is returned
+ * if the image does not exist.
+ ******************************************************************************/
+entry_point_info_t *bl31_plat_get_next_image_ep_info(uint32_t type)
+{
+	entry_point_info_t *next_image_info;
+
+	next_image_info = (type == NON_SECURE) ?
+			&bl33_image_ep_info : &bl32_image_ep_info;
+
+	/* None of the images on this platform can have 0x0 as the entrypoint */
+	if (next_image_info->pc)
+		return next_image_info;
+	else
+		return NULL;
+}
+
+/*******************************************************************************
+ * Perform any BL3-1 early platform setup. Here is an opportunity to copy
+ * parameters passed by the calling EL (S-EL1 in BL2 & S-EL3 in BL1) before they
+ * are lost (potentially). This needs to be done before the MMU is initialized
+ * so that the memory layout can be used while creating page tables.
+ * BL2 has flushed this information to memory, so we are guaranteed to pick up
+ * good data.
+ ******************************************************************************/
+void bl31_early_platform_setup(bl31_params_t *from_bl2,
+						 void *plat_params_from_bl2)
+{
+	mtk_bl_param_t *pmtk_bl_param_t = (mtk_bl_param_t *)from_bl2;
+	atf_arg_t_ptr teearg;
+
+	/*
+	 * Mediatek preloader(i.e, BL2) is in 32 bit state, high 32bits
+	 * of 64 bit GP registers are UNKNOWN if CPU warm reset from 32 bit
+	 * to 64 bit state. So we need to clear high 32bit,
+	 * which may be random value.
+	 */
+	pmtk_bl_param_t =
+	(mtk_bl_param_t *)((uint64_t)pmtk_bl_param_t & 0x00000000ffffffff);
+	plat_params_from_bl2 =
+	(void *)((uint64_t)plat_params_from_bl2 & 0x00000000ffffffff);
+
+	teearg  = (atf_arg_t_ptr)pmtk_bl_param_t->tee_info_addr;
+
+	console_init(teearg->atf_log_port, MT6795_UART_CLOCK, MT6795_BAUDRATE);
+	memcpy((void *)&gteearg, (void *)teearg, sizeof(atf_arg_t));
+
+	/* init system counter in ATF */
+	setup_syscnt();
+
+	VERBOSE("bl31_setup\n");
+
+	assert(from_bl2 != NULL);
+	/* Populate entry point information for BL3-2 and BL3-3 */
+	SET_PARAM_HEAD(&bl32_image_ep_info,
+				PARAM_EP,
+				VERSION_1,
+				0);
+	SET_SECURITY_STATE(bl32_image_ep_info.h.attr, SECURE);
+	bl32_image_ep_info.pc = BL32_BASE;
+	bl32_image_ep_info.spsr = plat_get_spsr_for_bl32_entry();
+
+	SET_PARAM_HEAD(&bl33_image_ep_info,
+				PARAM_EP,
+				VERSION_1,
+				0);
+	/*
+	 * Tell BL3-1 where the non-trusted software image
+	 * is located and the entry state information
+	 */
+	/* BL33_START_ADDRESS */
+	bl33_image_ep_info.pc = pmtk_bl_param_t->bl33_start_addr;
+	bl33_image_ep_info.spsr = plat_get_spsr_for_bl33_entry();
+	bl33_image_ep_info.args.arg4 =  pmtk_bl_param_t->bootarg_loc;
+	bl33_image_ep_info.args.arg5 =  pmtk_bl_param_t->bootarg_size;
+	SET_SECURITY_STATE(bl33_image_ep_info.h.attr, NON_SECURE);
```

Comment:
OK, now I understand. This is an unusual boot flow. I presume this is because your "LK and load linux boot image" executes at a different register width than the Linux kernel? Is this a temporary measure until you've ported your linux boot image to AArch64, or do you have a need to support both AArch32 and AArch64 kernels?

This is another area where we might need to consider more architected software support.